src/examples/int65.ts

```diff
@@ -0,0 +1,160 @@
+// TODO: review and potentially make this snarkyjs' Int64
+
+import { Field, UInt64, prop, CircuitValue, Circuit, isReady } from 'snarkyjs';
+
+export { Int65 };
+
+class Int65 extends CircuitValue {
+  // * in the range [-2^64+1, 2^64-1], unlike a normal int64
+  // * under- and overflowing is disallowed, similar to UInt64, unlike a normal int64
+
+  @prop magnitude: Field; // absolute value, restricted like UInt64
+  @prop sign: Field; // +/- 1
+
+  constructor(magnitude: Field, sign: Field) {
+    super();
+    this.magnitude = magnitude;
+    this.sign = sign;
+  }
+
+  static fromFieldUnchecked(x: Field) {
+    let MINUS_ONE = Field.one.neg();
+    let FIELD_ORDER = BigInt(MINUS_ONE.toString()) + 1n;
+    let TWO64 = 1n << 64n;
+    let xBigInt = BigInt(x.toString());
+    let isValidPositive = xBigInt < TWO64; // covers {0,...,2^64 - 1}
+    let isValidNegative = FIELD_ORDER - xBigInt < TWO64; // {-2^64 + 1,...,-1}
+    if (!isValidPositive && !isValidNegative)
+      throw Error(
+        `Int64.fromField expected a value between (-2^64, 2^64), got ${x}`
+      );
+    let magnitude = Field(isValidPositive ? x.toString() : x.neg().toString());
+    let sign = isValidPositive ? Field.one : MINUS_ONE;
+    return new Int65(magnitude, sign);
+  }
+
+  static fromUnsigned(x: UInt64) {
+    return new Int65(x.value, Field.one);
+  }
+
+  static fromNumber(x: number) {
+    return Int65.fromFieldUnchecked(Field(x));
+  }
+  static fromString(x: string) {
+    return Int65.fromFieldUnchecked(Field(x));
+  }
+  static fromBigInt(x: bigint) {
+    let xField = x < 0n ? Field((-x).toString()).neg() : Field(x.toString());
+    return Int65.fromFieldUnchecked(xField);
+  }
+
+  toString() {
+    let abs = this.magnitude.toString();
+    let sgn = this.sign.equals(Field.one).toBoolean() || abs === '0' ? '' : '-';
+    return sgn + abs;
+  }
+
+  isConstant() {
+    return this.magnitude.isConstant() && this.sign.isConstant();
+  }
+
+  // --- circuit-compatible operations below ---
+  // the assumption here is that all Int65 values that appear in a circuit are already checked as valid
+  // this is because Circuit.witness calls .check
```

Comment:
oo I didn't know this, where is this defined?

Reply:
It's here! https://github.com/MinaProtocol/mina/blob/5a1640263e9fcbec000d8e5709d5c3c5b0773ff5/src/lib/snarky_js_bindings/lib/snarky_js_bindings_lib.ml#L1323